### ISSUE TITLE
PWA for mobile

### ISSUE THREAD

**Issue by nolyoi** [SysLogica/w98-mono#8]

**Comment by nolyoi** [comment]
Do we want to do the idea of a mobile OS as it's own separate site?

**Comment by CanerKocak** [comment]
yo so like mobile.windoge98.com and desktop.windoge98.com gonna generate diff principals for users n stuff 😕 

 we could just use app.windoge98.com and then css to handle the diff layouts for mobile n desktop, this way we can:

- keep auth & principals same across all devices, so users have same account when logging in on phone or pc
- less code duplication / ship faster with one codebase also?
- way more efficient thats for sure
- im lazy and this is easy way to get mobile that actually isnt even that bad of a choice 🤙

**Comment by nolyoi** [comment]
> yo so like mobile.windoge98.com and desktop.windoge98.com gonna generate diff principals for users n stuff 😕
> 
> we could just use app.windoge98.com and then css to handle the diff layouts for mobile n desktop, this way we can:
> 
> * keep auth & principals same across all devices, so users have same account when logging in on phone or pc
> * less code duplication / ship faster with one codebase also?
> * way more efficient thats for sure
> * im lazy and this is easy way to get mobile that actually isnt even that bad of a choice 🤙

No, you can set up the ic-assets json to use the same principal for X, Y, Z domains. thats how ads.windoge98.com and dogverts.windoge.com and the canister url all gen the same principal

**Comment by nolyoi** [comment]
I think another potential solution would be building a separate canister for user management, I think thats what OC does.

Heres some gpt.

# I need a canister that generates a wallet/principal that can be used across my different canisters
ChatGPT
Internet Identity Overview
Internet Identity is a blockchain authentication method for the Internet Computer that uses WebAuthn. When a user logs in using Internet Identity, they are authenticated using their device (like a YubiKey, biometric device, or similar) and the Internet Identity service returns a user principal. This principal is a unique identifier for that user which does not change and can be used to identify them across different canisters on the Internet Computer.

Building a Canister for User Management
Since you need a system that allows users to authenticate once and have their identity be recognizable across multiple canisters, you'd likely want to create a user management canister. This canister would interface with the Internet Identity canister to authenticate users and then provide a means to use that authenticated identity across your ecosystem of canisters.

1. Authenticate with Internet Identity
First, integrate your user management canister with Internet Identity to handle authentication. You can store a mapping of the Internet Identity's user principal to a custom user profile that you manage.

2. Generate and Manage Custom Profiles
Upon successful authentication, you can generate a custom profile or wallet associated with the user's principal. You can then use this profile to manage additional user data and permissions across your canisters.

3. Creating a Custom Principal or Wallet
While you cannot directly create new principal identifiers (as they are system-generated), you can simulate this by linking any number of your internal identifiers or accounts to the principal obtained from Internet Identity. This setup can effectively act as a "wallet" that spans across your dApps.

4. Use Across Canisters
You can allow other canisters to query your user management canister to retrieve user information based on the principal provided by Internet Identity. This ensures that user identity is consistent and accessible across all your applications.

Example in Motoko
Here is a basic structure of what the user management canister code might look like in Motoko:

```motoko
import Principal "mo:base/Principal";

actor UserManagement {
    private stable var users : [(Principal, User)] = [];

    public type User = {
        id : Principal;
        data : UserData;
    };
    public type UserData = {
        username : Text;
        email : Text;
    };

    // Function to register or update a user after successful authentication with Internet Identity
    public func registerOrUpdateUser(authenticatedPrincipal : Principal, userData : UserData) : async Bool {
        let existingUser = Array.find(users, func ((p, _)) { p == authenticatedPrincipal });
        switch (existingUser) {
            case (null) {
                // New user
                users := Array.append(users, (authenticatedPrincipal, { id = authenticatedPrincipal; data = userData }));
            };
            case (?(_, user)) {
                // Update existing user
                users := Array.replace(users, func ((p, _)) { p == authenticatedPrincipal }, (authenticatedPrincipal, { id = authenticatedPrincipal; data = userData }));
            };
        };
        return true;
    }

    // Function to get user data by principal
    public func getUserData(principal : Principal) : async ?UserData {
        let user = Array.find(users, func ((p, _)) { p == principal });
        switch (user) {
            case (null) { null };
            case (?(_, userData)) { ?userData.data };
        };
    }
}
```

Integration Steps
Authentication: Handle the redirection and callback from Internet Identity.
Data Handling: On successful authentication, update or register user data in your user management canister.
Data Access: Provide interfaces for other canisters to access or verify user data.
This structure helps maintain a consistent user identity across multiple dApps by managing additional user data linked to the unique principal provided by Internet Identity. This approach keeps the principal consistent across different services, ensuring seamless user experience and integration.


# now how do i use that principal from other dapps?
ChatGPT
Using the principal retrieved from one dApp in other dApps in the DFINITY Internet Computer environment involves securely sharing and verifying the principal across different canisters. You'll generally achieve this by making inter-canister calls, where one canister (your user management or authentication canister) confirms the identity or shares relevant user details with another canister based on the principal.

Here are the steps and methods to facilitate this cross-canister interaction:

Step 1: Expose a Verification or Data Sharing Interface in Your User Management Canister
Your user management canister should have public methods that other canisters can call to verify a user's identity or retrieve user-related data. Let’s expand the example from the previous discussion to include such a method:

```motoko
actor UserManagement {
    ...
    // Public method to verify user principal and get basic data
    public query func verifyUser(principal : Principal) : async ?UserData {
        let user = Array.find(users, func ((p, _)) { p == principal });
        switch (user) {
            case (null) { null };
            case (?(_, userData)) { ?userData.data };
        };
    }
}
```
Step 2: Call This Method from Other dApps
Other dApps will need the principal of the user (obtained through the Internet Identity service) and the ability to make inter-canister calls to UserManagement. This is typically done using the IC’s actor model, where you create an actor instance pointing to the UserManagement canister from another canister and call its methods.

Example of another dApp querying `UserManagement`:
```motoko
import UserManagement "canister:user_management";
import Principal "mo:base/Principal";
import Actor "mo:base/Actor";

actor {
    private let userManager = actor("user-management-canister-id" : Principal) : UserManagement.UserManagement;

    // Function to get user data from UserManagement canister
    public query func getUserInfo(userPrincipal : Principal) : async ?UserManagement.UserData {
        return await userManager.verifyUser(userPrincipal);
    }
}
```
Step 3: Manage Permissions and Security
When exposing user data across canisters, consider the security and privacy implications:

Permissions: Make sure only authorized canisters can query sensitive user data. This might involve a list of allowed canisters or checking caller identities.
Data Minimization: Only expose the necessary amount of data required for the functionality of the other dApps.
Step 4: Test and Debug
Testing inter-canister calls is crucial:

Ensure that the canisters are deployed on the same network (local or mainnet).
Verify that the principal passed between canisters is not altered.
Use the DFINITY SDK’s local development environment to simulate calls and handle errors.
Step 5: Deploy and Monitor
Once tested, deploy your canisters and monitor their interactions. Regularly review your canister's logs and performance to ensure they are functioning as expected without security breaches or data leaks.

Additional Considerations
When implementing this system, think about scalability and efficiency. Inter-canister calls, especially those involving user data, need to be managed carefully to avoid bottlenecks or excessive query costs.

By following these steps, you can effectively use a principal authenticated via one dApp across other dApps on the Internet Computer, ensuring a seamless user experience while maintaining robust security practices.

**Comment by nolyoi** [comment]
Here's OC's setup: https://oc.app/architecture

![image](https://github.com/SysLogica/w98-mono/assets/59351042/b2ddffe8-43e5-42d3-bb9d-a2ec767e1cbe)
